File: csharp/grade-school/GradeSchool.cs
using System;
using System.Collections.Generic;
using System.Collections.Immutable;
using System.Linq;

public struct Student : IEquatable<Student>, IComparable<Student>
{
    public string Name;
    public int Grade;

    public Student(string name, int grade) : this()
    {
        Name = name;
        Grade = grade;
    }

    public int CompareTo(Student other) 
        => Grade == other.Grade 
        ? Name.CompareTo(other.Name) 
        : Grade - other.Grade;

    public bool Equals(Student other) => Name == other.Name;
}

public class GradeSchool
{
    private List<Student> _students { get; } = new();

    public bool Add(string name, int grade)
    {
        Student student = new Student(name, grade);
        if (_students.Contains(student))
        {
            return false;
        }
        _students.Add(student);
        return true;
    }

    public IEnumerable<string> Roster()
        => _students.Order().Select(s => s.Name);

    public IEnumerable<string> Grade(int grade) 
        => _students.Where(s => s.Grade == grade).Select(s => s.Name).Order();
}

File: nim/grade-school/grade_school.nim
import sequtils

type
  Student* = object
    name*: string
    grade*: int

  School* = object
    students*: seq[Student]

proc roster*(school: School): seq[string] =
  ## Returns the names of every student in the `school`, sorted by grade then name.
  discard



proc addStudent*(school: var School, name: string, grade: int) =
  ## Adds a student with `name` and `grade` to the `school`.
  ##
  ## Raises a `ValueError` if `school` already contains a student named `name`.
  school.students.add(Student(name: name, grade: grade))

proc grade*(school: School, grade: int): seq[string] =
  ## Returns the names of the students in the given `school` and `grade`, in
  ## alphabetical order.
  discard
